### PR TITLE
Aggressively push end-to-end test packages to minimize test run time

### PR DESCRIPTION
**Pull request by joelverhagen**
If end-to-end tests need N packages total and the first test is running, push all N packages instead of just the one that the test needs. This means they will all start flowing through V3 at the same time and subsequent tests won't have to wait long for the package to be available.

### CODE REVIEW

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs [hunk 1]
```diff
@@ -11,108 +11,178 @@
...
-                if (pushedPackage.Pushed)
+                var packageTypes = GetPackageTypes(requestedPackageType);
+
+                // Push all of the package types that have not been pushed yet.
+                var pushTasks = new Dictionary<PackageType, Task<Package>>();
+                foreach (var packageType in packageTypes)
                 {
-                    logger.WriteLine($"Package {pushedPackage} has already been pushed.");
-                    return pushedPackage.Package;
+                    if (!_packages.TryGetValue(packageType, out Package package))
```
Comment (scottbommarito) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: Not using `_packagesLock` here? I see `GetPushedPackage` uses it on read.

Can we just use `GetPushedPackage` and check if it returns null?

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs [hunk 2]
```diff
@@ -11,108 +11,178 @@
...
+            _pushLock = new SemaphoreSlim(initialCount: 1);
+            _packages = new Dictionary<PackageType, Package>();
         }
         
-        public async Task<Package> PushAsync(PackageType packageType, ITestOutputHelper logger)
+        public async Task<Package> PushAsync(PackageType requestedPackageType, ITestOutputHelper logger)
         {
-            PushedPackage pushedPackage;
-            lock (_packagesLock)
+            var pushedPackage = GetPushedPackage(requestedPackageType, logger);
```
Comment (scottbommarito) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: could reduce this method to `return GetPushedPackage(requestedPackageType, logger) ?? await PushUnpushedPackagesAsync(requestedPackageType, logger);`

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs [hunk 3]
```diff
@@ -11,108 +11,178 @@
...
+                    {
+                        pushTasks.Add(packageType, UncachedPushAsync(requestedPackageType, packageType, logger));
+                    }
                 }
 
-                using (var packageStream = new MemoryStream(pushedPackage.Package.NupkgBytes.ToArray()))
+                await Task.WhenAll(pushTasks.Values);
+
+                // Add the pushed packages to the cache.
+                lock (_packagesLock)
```
Comment (scottbommarito) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: You could remove the code inside this `lock` if you added `_packages[requestedPackageType] = package` to `UncachedPushAsync`.

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs [hunk 4]
```diff
@@ -11,108 +11,178 @@
...
-            if (pushedPackage.Pushed)
-            {
-                logger.WriteLine($"Package {pushedPackage} push has already been pushed.");
-                return pushedPackage.Package;
-            }
+            return await PushUnpushedPackagesAsync(requestedPackageType, logger);
+        }
 
-            var acquired = await pushedPackage.PushLock.WaitAsync(0);
+        private async Task<Package> PushUnpushedPackagesAsync(PackageType requestedPackageType, ITestOutputHelper logger)
```
Comment (scottbommarito) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: I think this function would read a lot easier if
- You had a function that checked if a single package was cached, and if not, pushed it and added it to the cache.
- This function called that function for all `PackageType`s and then returned `GetPushedPackage(requestedPackageType)`.

All the locks and caching could be handled inside the first method. This method just needs to make sure all packages are pushed and then check the cache.

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs [hunk 5]
```diff
@@ -11,108 +11,178 @@
...
-                PushLock = new SemaphoreSlim(1);
-                Package = package;
-                Pushed = false;
-            }
+        }
 
-            public SemaphoreSlim PushLock { get; }
-            public Package Package { get; }
-            public bool Pushed { get; private set; }
+        private IEnumerable<PackageType> GetPackageTypes(PackageType requestedPackageType)
```
Comment (scottbommarito) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: There is currently no visibility here if `requestedPackageType` is one of the `SemVer2PackageTypes`. I think this should throw or at least log if that is the case.

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs (hunk 2 above)
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: That's a long line, I think this is easier to read.

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs [hunk 6]
```diff
@@ -11,108 +11,178 @@
...
-                logger.WriteLine($"Package {pushedPackage} push has already been pushed.");
-                return pushedPackage.Package;
-            }
+            return await PushUnpushedPackagesAsync(requestedPackageType, logger);
+        }
 
-            var acquired = await pushedPackage.PushLock.WaitAsync(0);
+        private async Task<Package> PushUnpushedPackagesAsync(PackageType requestedPackageType, ITestOutputHelper logger)
+        {
+            var acquired = await _pushLock.WaitAsync(0);
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: Never seen it before but I like it. It gives a more detailed test output.

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs [hunk 7]
```diff
@@ -11,108 +11,178 @@
...
             try
             {
                 if (!acquired)
                 {
-                    logger.WriteLine($"Another test is already pushing {pushedPackage}. Waiting.");
-                    await pushedPackage.PushLock.WaitAsync();
-                    acquired = true;
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: Yes, the initial `WaitAsync` will either set acquired to `true` (in which case this scope is not execute`) or to `false` in which case this scope is executed and `acquired` is still set to `true`.

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs [hunk 8]
```diff
@@ -11,108 +11,178 @@
...
+                    logger.WriteLine("Another test is already pushing the packages. Waiting.");
+                    await _pushLock.WaitAsync();
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: User's of `SemaphoreSlim` must assume that either `WaitAsync` throws and the lock is not actually acquired or it returns a correct `acquired` boolean (either `true` or `false`). If this assumption is made, then we are safe.

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs (hunk 1 above)
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: Yeah, good idea.

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs (hunk 5 above)
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: I wanted to be "correct". Enums can have integer values that do not correspond to a name.

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs (hunk 4 above)
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: The problem with this is that you will not be pushing in parallel.

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs (hunk 7 above)
Comment (ryuyu) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: I'm a little confused. Where in this scope is acquired set to true? the line setting it to true explicitly appears to have been removed with your changeset?

Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: Discussed offline. Added the `acquired = true` back. Thanks guys.

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs (hunk 3 above)
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: Fixed.

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs (hunk 5 above)
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: Fixed.

**Comment by joelverhagen** [comment]
Thought about this more. Added unit tests. I think this is a good precedent to set albeit not totally necessary in this case.

**Pull request merged**